La/prove edwards decompress

astefano opened the pull request:
<!-- Please ensure that your PR includes the following, as needed -->

- [x] Source code modifications **highlighted and justified**
- [x] Proof cheats (`assume(false)`, `sorry`, etc.) **highlighted and justified**

This PR proves Edwards::step_1 and Edwards::decompress. Uses 3 "axioms":
- `axiom_sqrt_m1_squared` — i² = -1 (mod p)
- `axiom_sqrt_m1_not_square` — i is not a square (Euler's criterion)
- `axiom_neg_sqrt_m1_not_square` — -i is not a square

### **Specification changes:**
- decompress has a precondition: ` is_valid_compressed_edwards_y(&self.0)`
  Justification: The compressed representation must have a valid sign bit. This is automatically satisfied for points from compress(). For external bytes, callers must validate.
- step_1: extended postconditions needed for the proof of decompress
- step_2: relaxed precondition to  align with what sqrt_ratio_i actually produces.

- conditional_negate_field_element: we relaxed the precondition from 51 to 52 to match mul/square output which are used in decompress

- FieldElement::sqrt_ratio_i has bounds as preconditions and `(spec_field_element(&result.1) % p()) % 2 == 0` and `spec_field_element(&result.1) < p()` as postconditions. 

### Dependency graphs:

edwards.rs::step_1()
    │
    └──► step1_lemmas.rs
              │
              ├──► lemma_step1_case_analysis (main entry)
              │         │
              │         ├──► lemma_sqrt_ratio_success_means_valid_y
              │         ├──► lemma_sqrt_ratio_failure_means_invalid_y
              │         └──► lemma_sqrt_ratio_implies_on_curve
              │
              └──► sqrt_ratio_lemmas.rs
                        │
                        ├──► lemma_no_square_root_when_times_i
                        │         │
                        │         ├──► **axiom_sqrt_m1_not_square**  
                        │         ├──► **axiom_neg_sqrt_m1_not_square**
                        │         └──► lemma_algebraic_chain_base
                        │
                        ├──► lemma_flipped_sign_becomes_correct
                        │         └──► lemma_multiply_by_i_flips_sign
                        │
                        └──► field_algebra_lemmas.rs
                                  │
                                  ├──► lemma_quotient_of_squares
                                  ├──► lemma_solve_for_left_factor
                                  ├──► lemma_inv_of_product
                                  ├──► lemma_double_negation
                                  └──► lemma_field_mul_comm

(see https://beneficial-ai-foundation.github.io/scip-callgraph/?source=edwards%3A%3Astep_1&exclude=*_mul*%2C+*_mod*%2C+pow*%2C+p_*&depth=4&hidden=square%2Csqrt_ratio_i%2Cfrom_bytes%2Cas_bytes for the full graph)                                  
                                  
decompress() 
│
├── step_1() → step1_lemmas.rs
│
└── lemma_decompress_valid_branch 
    │
    ├──► lemma_negation_preserves_curve
    │         ├──► lemma_neg_square_eq              
    │         └──► lemma_square_mod_noop         
    │
    ├──► lemma_decompress_produces_valid_point 
    │         ├──► lemma_field_inv_one              
    │         └──► lemma_square_mod_noop           
    │
    ├──► lemma_sign_bit_one_implies_x_nonzero 
    │
    └──► lemma_decompress_field_element_sign_bit                         

(see https://beneficial-ai-foundation.github.io/scip-callgraph/?source=edwards%3A%3Adecompress&exclude=p_*%2C*_mod*%2Cpow*%2C*_mul*%2C&depth=5&hidden=lemma_small_mod%2Cchoice_into%2Cstep_1%2Cstep_2 for the full graph)

closes https://github.com/Beneficial-AI-Foundation/dalek-lite/issues/399 
closes https://github.com/Beneficial-AI-Foundation/dalek-lite/issues/494

astefano commented:
There are a few lemmas that are used for the proof of FieldElement::sqrt_ratio_i(), so not necessary for the proof of decompress; if you prefer that we don't change FieldElement::sqrt_ratio_i() in this PR, i'll take them out and keep them locally in case they'll be needed later.

astefano commented:
> > There are a few lemmas that are used for the proof of FieldElement::sqrt_ratio_i(), so not necessary for the proof of decompress; if you prefer that we don't change FieldElement::sqrt_ratio_i() in this PR, i'll take them out and keep them locally in case they'll be needed later.
> 
> my 2c: this PR is quite large, any slim-down will make it easier to read

ok, done.

astefano commented on curve25519-dalek/src/edwards.rs:
```diff
@@ -228,11 +242,18 @@ impl CompressedEdwardsY {
+        requires
+            is_valid_compressed_edwards_y(&self.0),
         ensures
             math_is_valid_y_coordinate(spec_field_element_from_bytes(&self.0))
```
yes, i was hesitating, what do you think? 
if decompress is supposed to only accept valid points, we'd like to check that. if we have it valid_point -> ensures, we'd miss that decompress had as input an invalid point.

astefano commented on curve25519-dalek/src/lemmas/common_lemmas/mod.rs:
```diff
@@ -12,4 +12,8 @@ pub mod pow_lemmas;
 
 pub mod shift_lemmas;
 
+pub mod sqrt_ratio_lemmas;
```
see https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/576#issuecomment-3650310826

astefano commented on curve25519-dalek/src/lemmas/common_lemmas/sqrt_ratio_lemmas.rs:
```diff
@@ -0,0 +1,977 @@
+//! Lemmas for sqrt_ratio_i algorithm verification
```
We need them for the proof of step_1 (previously, i removed only one lemma i was using for the proof of `sqrt_ratio_i` as we decided we leave `sqrt_ratio_i` as is on main; i kept though, the change in pre/postconditions as noted in the PR description).

astefano replied:
(i agree the comment is misleading; so far, the file contains lemmas for proving decompress; later on, it will contain lemmas for FieldElement::sqrt_ratio_i as well.)

astefano commented on curve25519-dalek/src/lemmas/edwards_lemmas/decompress_lemmas.rs:
```diff
@@ -0,0 +1,520 @@
+    lemma_edwards_d_limbs_bounded_54();
+}
+
+/// Main Lemma: decompress produces a valid Edwards point
```
right, thanks, done: https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/576/commits/3bfebd4b4265b4911f8aaaf2b1898ddd45333713

astefano commented on curve25519-dalek/src/lemmas/edwards_lemmas/step1_lemmas.rs:
```diff
@@ -0,0 +1,600 @@
+pub proof fn lemma_u_zero_implies_identity_point(y: nat, u: nat)
+    requires
+        u == math_field_sub(math_field_square(y), 1),
+        u == 0,
```
right, done: https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/576/commits/1f4f129e0f1846bf18ee17e1ae46d99b6386ad3e

astefano commented on curve25519-dalek/src/lemmas/edwards_lemmas/step1_lemmas.rs:
```diff
@@ -0,0 +1,600 @@
+///   u64_5_as_nat(r.limbs) % p() == pow(u64_5_as_nat(self.limbs) as int, 2) as nat % p()
+///
+/// Combined with the fact that pow(n, 2) = n * n and modular arithmetic properties.
+pub proof fn lemma_square_matches_math_field_square(y_raw: nat, y2_raw: nat)
```
right, thanks, done: https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/576/commits/ea26c4d4242808131e0516df20a7efd98333b166

astefano commented:
> Lots of beneficial foundational lemmas in this PR !

as always, i appreciate your sense of humour :)

astefano commented on curve25519-dalek/src/lemmas/edwards_lemmas/decompress_lemmas.rs:
```diff
@@ -0,0 +1,520 @@
+    ensures
+        ((x_after_negate % p()) % 2) as u8 == (repr_byte_31 >> 7),
+{
+    let sign_bit = repr_byte_31 >> 7;
```
right, thanks, done: https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/576/commits/891f914155e0d36268356e56e77a3a3910109bd2

astefano commented on curve25519-dalek/src/lemmas/edwards_lemmas/step1_lemmas.rs:
```diff
@@ -0,0 +1,600 @@
+//! - Y coordinate from compressed bytes
+//! - Z = 1
+//! - sqrt_ratio_i(u, v) where u = Y² - 1 and v = d·Y² + 1
+//!
```
done: https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/576/commits/cc7bf7696209655de240d22284bee4175078eea1

sergiubur commented on curve25519-dalek/src/lemmas/edwards_lemmas/step1_lemmas.rs:
```diff
@@ -0,0 +1,600 @@
+            let y2 = math_field_square(y);
+            u_math == math_field_sub(y2, 1) && v_math == math_field_add(math_field_mul(d, y2), 1)
+        }),
+        // sqrt_ratio_i postconditions (from ensures clause)
```
Seems that not all sqrt_ratio_i postconditions are in the spec function - ?

sergiubur commented on curve25519-dalek/src/edwards.rs:
```diff
@@ -311,71 +336,142 @@ mod decompress {
+
+            //prove the last 2 postconditions for step_1
+            // From sqrt_ratio_i ensures: non-negative root, bounded
+            assert((spec_field_element(&X) % p()) % 2 == 0);
```
now these are included in the postcondition spec function ? Since we just called sqrt_ratio_i, I wonder if we can move up the corresponding assert on postcondition here - and whether we need it at all.

astefano replied:
right, thanks, done: https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/576/commits/2a2d3da2f0bc00f8ea7a41cf5c0d66298127a7f2

sergiubur commented on curve25519-dalek/src/lemmas/common_lemmas/sqrt_ratio_lemmas.rs:
```diff
@@ -0,0 +1,972 @@
+/// Used in: lemma_flipped_sign_becomes_correct (below)
+pub proof fn lemma_multiply_by_i_flips_sign(r: nat)
+    ensures
+        ((r * spec_sqrt_m1()) % p() * (r * spec_sqrt_m1()) % p()) % p() == ((p() as int - ((r * r)
```
does it work if we use pow((r * spec_sqrt_m1()) % p(),2) instead ? Seems simpler to parse and would match the informal description. 

actually, we also have math_field_mul, where we wouldn't need to use %p. That would be also make it more readable I think.

sergiubur commented on curve25519-dalek/src/lemmas/common_lemmas/sqrt_ratio_lemmas.rs:
```diff
@@ -0,0 +1,972 @@
+    };
+}
+
+/// Corollary: When r_squared_v = -u, we have (-u) · inv(i·u) = i
```
same, general fact, independent of r.

sergiubur commented on curve25519-dalek/src/lemmas/common_lemmas/sqrt_ratio_lemmas.rs:
```diff
@@ -0,0 +1,972 @@
+            lemma_small_mod(q2, the_p);
+            lemma_small_mod(i, the_p);
+        };
+        assert(is_square_mod_p(i));
```
assert(false); // contradiction 
?

sergiubur commented on curve25519-dalek/src/lemmas/edwards_lemmas/step1_lemmas.rs:
```diff
@@ -0,0 +1,550 @@
+        // Field element bounds (u_math and v_math are field elements, so < p)
+        x < p(),
+        u_math < p(),
+        v_math < p(),
```
these already implied by above spec_...  ?

sergiubur commented on curve25519-dalek/src/lemmas/field_lemmas/field_algebra_lemmas.rs:
````diff
@@ -0,0 +1,1002 @@
+///            = math_field_mul(x_raw % p, y_raw % p)   [definition of math_field_mul]
+/// ```
+///
+/// This is useful for converting from FieldElement51::mul postconditions
````
We can remove references to dalek.

sergiubur commented on curve25519-dalek/src/lemmas/field_lemmas/field_algebra_lemmas.rs:
````diff
@@ -0,0 +1,1002 @@
+/// Subtract b: a + 1 ≡ c - b (mod p)  ✓
+/// ```
+///
+/// Used in sqrt_ratio_implies_on_curve proof.
````
We can remove references to dalek

astefano replied:
done

sergiubur commented:
Noticing again how many useful generic lemmas are in this PR, we should consider making a separate arithmetic field library, potentially including the affine curve specs over that field.

sergiubur commented on curve25519-dalek/src/lemmas/field_lemmas/field_algebra_lemmas.rs:
```diff
@@ -0,0 +1,1002 @@
+
+    // Step 5: Chain: a % p = (a*b*inv_b) % p = (ab_mod * inv_b) % p = c * inv_b % p
+}
+
```
I think field arithmetic is undecidable, but we might make it LLM-decidable if LLM learn to write such lemmas for equalities that are needed case by case. Actually, this is a special case where we know the equality holds and only want to find the proof. That might be in some decidable but still untractable complexity class.

sergiubur commented on curve25519-dalek/src/specs/edwards_specs.rs:
```diff
@@ -369,6 +377,43 @@ pub open spec fn compressed_edwards_y_corresponds_to_edwards(
+/// 1. **Mathematical basis**: For points with x = 0 on the Edwards curve,
+///    the curve equation gives y² = 1, so y = ±1. These special points
+///    (the identity (0,1) and the point (0,-1)) have only one valid sign bit: 0.
+///
```
Maybe the mathematical justification is sufficient and we can remove points 2-4 ?

sergiubur commented on docs/proofs_for_decompress/decompress_math_proof_mapped_to_verus.md:
```diff
@@ -0,0 +1,656 @@
+| `axiom_sqrt_m1_not_square` | $i$ is not a quadratic residue | Euler criterion: $i^{(p-1)/2} = -1$ | `sqrt_ratio_lemmas.rs` |
+| `axiom_neg_sqrt_m1_not_square` | $-i$ is not a quadratic residue | Euler criterion: $(-i)^{(p-1)/2} = -1$ | `sqrt_ratio_lemmas.rs` |
+| `axiom_p_is_prime` | $p = 2^{255} - 19$ is prime | Well-known mathematical fact | `primality_specs.rs` |
+| `lemma_binomial_factorial_relation` | $\binom{n}{k} \cdot k! \cdot (n-k)! = n!$ | Standard combinatorial identity | `number_theory_lemmas.rs` |
```
did a quick search and actually haven't seen binomial factorial or binomial axiom used.

astefano replied:
right, neither could i, it was introduced in the PR for FLT https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/353/files#diff-1a899ab742611bcde0050452a66a79557f710ffe835c6272d1abd4976b6d2a89R417 (i think it's a good idea to rename lemmas with admit() as axioms, at least it would help keeping track of admits...)
the callees seem to be like this: https://beneficial-ai-foundation.github.io/scip-callgraph/?sink=lemma_binomial_factorial_relation&depth=10

i updated the md file.

astefano commented on curve25519-dalek/src/specs/field_specs.rs:
```diff
@@ -450,6 +450,43 @@ pub open spec fn is_sqrt_ratio_times_i(
+        &&
+    // When unsuccessful and v ≠ 0 and u ≠ 0: r² * v ≡ i*u (mod p)
+    ((!success && v != 0 && u != 0) ==> math_is_sqrt_ratio_times_i(u, v, r))
+}
```
right, now it's separated: https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/576/commits/9f35159a9c29352464121c968db4dfab237df385

astefano commented on curve25519-dalek/src/lemmas/common_lemmas/sqrt_ratio_lemmas.rs:
```diff
@@ -0,0 +1,972 @@
+            let r_prime = (r * spec_sqrt_m1()) % p();
+            (v * r_prime * r_prime) % p() == u % p()
+        }),
+{
```
right, done https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/576/commits/97dadbb44fc3257c0c0be38e8691096fe35ebecc

astefano commented on curve25519-dalek/src/lemmas/common_lemmas/sqrt_ratio_lemmas.rs:
```diff
@@ -0,0 +1,972 @@
+    };
+}
+
+/// Corollary: When r_squared_v = u, we have u · inv(i·u) = -i
```
right, done

astefano commented on curve25519-dalek/src/lemmas/common_lemmas/sqrt_ratio_lemmas.rs:
right, thanks :blush:
moved it in field_lemmas

astefano commented:
> Edwards lemmas done

Thanks Jure! I addressed all your comments in this round here https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/576/commits/e1966705cd1eedf2f9751f3983d6d1926702d246. If you see that i overlooked something, let me know.

Kukovec commented on curve25519-dalek/src/lemmas/field_lemmas/field_algebra_lemmas.rs:
```diff
@@ -0,0 +1,999 @@
+    assert(math_field_add(a, 1) == math_field_sub(c, b)) by {
+        // Small mod simplifications
+        lemma_small_mod(c, p);
+        lemma_small_mod(1nat, p);
```
```suggestion
        lemma_small_mod(1, p);
```

Kukovec commented on curve25519-dalek/src/lemmas/field_lemmas/field_algebra_lemmas.rs:
````diff
@@ -0,0 +1,999 @@
+/// ```
+pub proof fn lemma_solve_for_left_factor(a: nat, b: nat, c: nat)
+    requires
+        b % p() != 0,
````
I think this is also technically not required, because we define inv(0) = 0, but this is more of a remark than a request for removal. Maybe add a comment

Kukovec commented on curve25519-dalek/src/lemmas/field_lemmas/sqrt_m1_lemmas.rs:
```diff
@@ -0,0 +1,427 @@
+///         ≡ p - r² mod p     [representation of negation]
+///
+/// Used in: lemma_flipped_sign_becomes_correct
+pub proof fn lemma_multiply_by_i_flips_sign(r: nat)
```
`lemma_multiply_by_i_flips_square_sign` maybe?

Kukovec commented on curve25519-dalek/src/lemmas/field_lemmas/sqrt_ratio_lemmas.rs:
```diff
@@ -0,0 +1,605 @@
+    }
+}
+
+/// If v·r² = -u, then v·(r·i)² = u
```
Possible proof simplification?
``` 
v * r^2 = -u
i^2 = -1
v * r^2*i^2 = -1 * -u
v * (ri)^2 = u
```

astefano replied:
here's an attempt: https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/576/commits/777e115898529d4a8825acf74cd8490a3e9c2380

Kukovec commented on curve25519-dalek/src/specs/field_specs.rs:
```diff
@@ -450,6 +450,58 @@ pub open spec fn is_sqrt_ratio_times_i(
+/// Use this when working with mathematical values directly in lemmas.
+pub open spec fn math_is_sqrt_ratio(u: nat, v: nat, r: nat) -> bool {
+    (r * r * v) % p() == u
+}
```
I see this is already defined, one of my comments above mentions using it

Kukovec reviewed (approved):
LGTM (didn't read `.md`)

astefano commented on curve25519-dalek/src/lemmas/field_lemmas/sqrt_ratio_lemmas.rs:
```diff
@@ -0,0 +1,605 @@
+    requires
+// is_sqrt_ratio condition: (x * x * v) % p == u
+
+        (x * x * v) % p() == u,
```
right, done https://github.com/Beneficial-AI-Foundation/dalek-lite/pull/576/commits/6176241d0adbe2df1183cfb21b0b4ccba5e69116
